src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.ILLink.targets

```diff
@@ -0,0 +1,115 @@
+<!--
+***********************************************************************************************
+Microsoft.NET.ILLink.targets
+
+WARNING:  DO NOT MODIFY this file unless you are knowledgeable about MSBuild and have
+          created a backup copy.  Incorrect changes to this file will make it
+          impossible to load or build your projects from the command-line or the IDE.
+
+Copyright (c) .NET Foundation. All rights reserved.
+***********************************************************************************************
+-->
+<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+
+  <Import Project="Sdk.props" Sdk="ILLink.Tasks" />
+
+  <PropertyGroup>
+    <ILTransformDependsOn>$(ILTransformDependsOn);ILLink</ILTransformDependsOn>
+    <IntermediateLinkDir Condition=" '$(IntermediateLinkDir)' == '' ">$(IntermediateOutputPath)linked</IntermediateLinkDir>
+    <!-- Used to enable incremental build for the linker target. -->
+    <_LinkSemaphore>$(IntermediateOutputPath)Link.semaphore</_LinkSemaphore>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <LinkerRootDescriptors Include="$(LinkerRootDescriptors)" />
+  </ItemGroup>
+
+  <!--
+    ============================================================
+                     ILLink
+
+    Replace the files to be published with versions that have been
+    passed through the linker. Also prevent removed files from being
+    included in the generated deps.json.
+    ============================================================
+    -->
+  <Target Name="ILLink"
+          Condition=" '$(LinkDuringPublish)' == 'true' "
+          DependsOnTargets="RunILLink">
+
+    <!-- For now, use ResolvedFileToPublish as input/output. This
+         should go away in favor of a well-defined set of runtime
+         assemblies with https://github.com/dotnet/sdk/issues/3109. -->
+    <ItemGroup>
+      <_LinkedResolvedFileToPublish Include="@(_LinkedResolvedFileToPublishCandidates)" Condition="Exists('%(Identity)')" />
+      <ResolvedFileToPublish Remove="@(_ManagedAssembliesToLink)" />
+      <ResolvedFileToPublish Include="@(_LinkedResolvedFileToPublish)" />
+    </ItemGroup>
+
+    <!-- Remove assemblies from inputs to GenerateDepsFile. See
+         https://github.com/dotnet/sdk/pull/3086. -->
+    <ItemGroup>
+      <_RemovedManagedAssemblies Include="@(_ManagedAssembliesToLink)" Condition="!Exists('$(IntermediateLinkDir)/%(Filename)%(Extension)')" />
+
+      <ResolvedCompileFileDefinitions Remove="@(_RemovedManagedAssemblies)" />
+      <NativeCopyLocalItems Remove="@(_RemovedManagedAssemblies)" />
+      <ResourceCopyLocalItems Remove="@(_RemovedManagedAssemblies)" />
+      <RuntimeCopyLocalItems Remove="@(_RemovedManagedAssemblies)" />
+      <RuntimeTargetsCopyLocalItems Remove="@(_RemovedManagedAssemblies)" />
+      <RuntimePackAsset Remove="@(_RemovedManagedAssemblies)" />
```

Comment:
This is the kind of change that would be undesirable if we moved this targets file to the ILLink nuget package, because if the names here change, this would be breaking.
@nguerrera, how can we ensure that name changes here won't break other tools/SDKs that depend on this? CI tests good enough?

Reply:
if other sdks to depend on it is expected, it should be public and follow backward compact rules

Reply:
This is exactly why we are putting the targets portion in the SDK. The contract to the linker is pure, no global mutations. Yes, other tools might do this sort of thing, but we specifically cut defining a strict contract for such tools. The linker will be first class and any mutation of the build state will happen in dotnet/sdk. This + sufficient tests will ensure that we don't just break the linker as we used to when it was it's own nuget package. I want to make it possible to write these sort of things outside of the sdk, but it is a hard problem and specifically not one we are going to solve as part of this work.